### DISCUSSION TITLE
fork.c:(.text+0x124c): undefined reference to `bprint' - when building w/AppleTalk for OpenWrt

### DISCUSSION THREAD

**Discussion by APCCV** [Netatalk/netatalk#1694] [Q&A] [answered]
Hi,

I have a stable build without AppleTalk. Building on an Ubuntu variant, cross compiling for an ARM CPU (most builds for OpenWrt involve cross compiling).
If I enable it for build, process halts with error above.
Netatalk 4.0.2 or 4.0.3 - same behaviour.

I can see the source for it:
`libatalk/util/bprint.c`
Can also see it was built:
`libatalk/util/libutil.a.p/bprint.c.o`

Error:
`/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/bin/../lib/gcc/arm-openwrt-linux-muslgnueabi/13.3.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld.bfd: etc/afpd/libafpd.a.p/fork.c.o: in function 'write_fork.isra.0':`
```fork.c:(.text+0x124c): undefined reference to 'bprint'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Missing include?
Any suggestions on what to poke at? Are additional libraries required to enable AppleTalk?

**Comment by APCCV** [comment]
Link command:
`arm-openwrt-linux-muslgnueabi-gcc  -o etc/afpd/afpd etc/afpd/libafpd.a.p/afp_config.c.o etc/afpd/libafpd.a.p/afp_dsi.c.o etc/afpd/libafpd.a.p/afp_options.c.o etc/afpd/libafpd.a.p/afp_util.c.o etc/afpd/libafpd.a.p/afprun.c.o etc/afpd/libafpd.a.p/appl.c.o etc/afpd/libafpd.a.p/auth.c.o etc/afpd/libafpd.a.p/catsearch.c.o etc/afpd/libafpd.a.p/desktop.c.o etc/afpd/libafpd.a.p/dircache.c.o etc/afpd/libafpd.a.p/directory.c.o etc/afpd/libafpd.a.p/enumerate.c.o etc/afpd/libafpd.a.p/extattrs.c.o etc/afpd/libafpd.a.p/fce_api.c.o etc/afpd/libafpd.a.p/fce_util.c.o etc/afpd/libafpd.a.p/file.c.o etc/afpd/libafpd.a.p/filedir.c.o etc/afpd/libafpd.a.p/fork.c.o etc/afpd/libafpd.a.p/hash.c.o etc/afpd/libafpd.a.p/mangle.c.o etc/afpd/libafpd.a.p/messages.c.o etc/afpd/libafpd.a.p/ofork.c.o etc/afpd/libafpd.a.p/status.c.o etc/afpd/libafpd.a.p/switch.c.o etc/afpd/libafpd.a.p/uam.c.o etc/afpd/libafpd.a.p/unix.c.o etc/afpd/libafpd.a.p/volume.c.o etc/afpd/libafpd.a.p/main.c.o etc/afpd/libafpd.a.p/afp_asp.c.o -L/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/usr/lib -L/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/lib -Wl,--as-needed -Wl,--no-undefined -fuse-ld=bfd -Wl,-export-dynamic -Wl,--start-group -ldl -lm -fuse-ld=bfd -znow -zrelro '-Wl,-rpath,$ORIGIN/../../libatalk:/usr/lib' -Wl,-rpath-link,/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/netatalk-4.0.3/openwrt-build/libatalk libatalk/libatalk.so.19.0.0 -L/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/usr/lib -lgcrypt -Wl,--end-group -pthread`
(in case it helps)

**Comment by APCCV** [comment]
Seems `bprint()` is only built if `DEBUG` is set. Looking at ``libatalk/util/bprint.c":
```
#ifdef DEBUG

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <atalk/util.h>

static const char       hexdig[] = "0123456789ABCDEF";

#define BPXLEN  50
#define BPALEN  18

void bprint( data, len )
...
``` 

But many files in `libatalk/atp` have a `EBUG`. For example `libatalk/atp/atp_rsel.c`. Looks suspiciously as a typo.

The offending call in fork.c is not wrapped in `[D]EBUG`... which means that maybe this only works if I build with `DEBUG` on. I'll try that, and the `EBUG` undefined will mean none of those DEBUG lines of code will kick in. Let's see.

**Reply by rdmark** [reply]
Good catch! I think it's incorrect to have this code under a DEBUG flag. Out of curiosity, I went back and checked when it was added. What I found was a commit in 2005, very very early in the 3.0 development cycle. The `bprint()` symbol is only used in AppleTalk code, so I think it was part of the preparations for removing AppleTalk support. The 2.x code does *not* have it.

https://github.com/Netatalk/netatalk/commit/ecfc96169ab669b578e53fa8e13592934fe37788#diff-c2c91dad975ae17671ca4aeff37c12be0803eb8a0c05e5262e27122c24fc79f0

The `EBUG` flag is clearly just for one-off or transient debugging that prior maintainers have used, what you get from passing `-DEBUG` to the compiler (as opposed to `-DDEBUG`).

**Comment by APCCV** [comment]
Ha. Didn't expect this. Add `-Dwith-debug=true` to meson arguments and it ignores it. Config reports `    with-debug                 : false
`

**Reply by rdmark** [reply]
Compile time debug code has mostly been removed a long time ago. We have shifted to runtime debug modes (either debug log levels, or options to enable debug modes in afp.conf). I will take a task to remove the last remnants.

**Comment by APCCV** [comment]
Ok. Force bprint to compile... and we hit another.

```
[266/269] Linking target etc/papd/papd
FAILED: etc/papd/papd 
arm-openwrt-linux-muslgnueabi-gcc  -o etc/papd/papd etc/papd/papd.p/auth.c.o etc/papd/papd.p/comment.c.o etc/papd/papd.p/file.c.o etc/papd/papd.p/headers.c.o etc/papd/papd.p/lp.c.o etc/papd/papd.p/magics.c.o etc/papd/papd.p/main.c.o etc/papd/papd.p/ppd.c.o etc/papd/papd.p/print_cups.c.o etc/papd/papd.p/printcap.c.o etc/papd/papd.p/queries.c.o etc/papd/papd.p/session.c.o etc/papd/papd.p/uam.c.o -L/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/usr/lib -L/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/lib -Wl,--as-needed -Wl,--no-undefined -fuse-ld=bfd -Wl,-export-dynamic -Wl,--start-group -ldl -lm -fuse-ld=bfd -znow -zrelro '-Wl,-rpath,$ORIGIN/../../libatalk:/usr/lib' -Wl,-rpath-link,/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/netatalk-4.0.3/openwrt-build/libatalk libatalk/libatalk.so.19.0.0 -Wl,--end-group
/home/work/netatalk/openwrt-sdk-ipq806x-generic_gcc-13.3.0_musl_eabi.Linux-x86_64/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-13.3.0_musl_eabi/bin/../lib/gcc/arm-openwrt-linux-muslgnueabi/13.3.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld.bfd: etc/papd/papd.p/lp.c.o: in function `lp_conn_inet':
lp.c:(.text+0x1c14): undefined reference to `rresvport'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
rresvport now. That's a `#ifdef MUSL`.

But that's the last one. Eliminating that `#ifdel` results in a successful link.

TO SUMMARIZE:
- link of afpd fails because bprint() is not compiled, because it's only compiled in debug mode.
- with-debug doesn't seem to work... forcing bprint() to build by removing the debug condition solves afpd
- link of papd fails because rresvport() is not compiled, because it's only compiled if MUSL is defined.
- remove the MUSL condition and papd links OK
- package builds after that

I don't have any AppleTalk devices so I can't test if if works or not.

I guess I better stop posting and let you take a look.

**Reply by rdmark** [reply]
That's weird, so your host machine is not detected as a musl system even though it is a musl system. Maybe it's better to check for a `rresvport()` implementation instead.

**Comment by rdmark** [comment, marked as answer]
@APCCV The `main` branch of netatalk should have fixes that hopefully resolves all of the issues above. Please give it a go when you have the chance. I'm looking forward to your feedback.

**Reply by APCCV** [reply]
I see the changes.
I have not developed the skill to use the OpenWrt SDK without a release (ie, from the HEAD of the main branch). It can be done, probably easily, but require changes to the already iffy OpenWrt package Makefile.
I'm pushing for 4.0.3 with relatively equivalent patches to be merged to the repo instead.
Once 4.0.4 is out I'll remove the patches and post feedback.
I need to find a way to test the AppleTalk piece. Maybe I can get two OpenWrt installs AppleTalking to each other.

**Reply by APCCV** [reply]
Packaged 4.0.4 for OpenWrt and issues above were no longer present. Thanks!

**Reply by rdmark** [reply]
Wonderful! Is your OpenWrt package published somewhere?

**Reply by APCCV** [reply]
Main entry point: https://downloads.openwrt.org/

It's being built for all current architectures (after a problem of my own making). Here are two:
https://downloads.openwrt.org/snapshots/packages/arm_cortex-a7/packages/netatalk_4.0.4-r1_arm_cortex-a7.ipk
https://downloads.openwrt.org/snapshots/packages/aarch64_generic/packages/netatalk_4.0.4-r1_aarch64_generic.ipk

That's for snapshot (bleeding edge). v3.2.5 was there before and now it's been replaced with v4.0.4. This build includes afpd and AppleTalk.

Release 24.10 is being branched and netatalk v4 made it in. Once it is available (starting with RC1) netatalk will be released under 24.10.

For a future release I'll try getting two or three variants built: a bare bones with afps only, apfs+atalkd, afps+kerberos... I'll try think what makes sense. Storage footprint vs functionality is the key consideration. 

Spotlight is not workable because sparql is not available. iconv... I've tried to make it work but have build issues (if I can I'll make a PR for it) and I have to understand if there is any real gain from using it. afpstats can be built but results in a huge storage footprint with all the libraries it needs. Zeroconf... need to look at (I see the value in it). Cracklib... tough to make a case for it. Other stuff... need to look at. Suggestions welcome.

I'll poke at it as time allows.

**Reply by rdmark** [reply]
Zeroconf is definitely very valuable. I wouldn't worry about Cracklib. Without iconv, you may lose support for non-Western European Mac codepages (Greek, Hebrew, Korean, Cyrillic and so on.)

Anyways, I listed you as Netatalk packager in CONTRIBUTORS, as "APCCV" (your display name here.) Please let me know if you prefer credits under a different name.
